### PR TITLE
[astro] Fix returning wrong sun phase name

### PR DESCRIPTION
**Pull request by lsiepel**
While troubleshooting [issue 7642](https://github.com/openhab/openhab-addons/issues/7642) first a minor bug was found in the calculation of sun phases.

In `Sun.getSunInfo` method for each phase (sunrise, sunset etc.) a range (start/end datetime) where that phase is valid for is calculated. At the end of all the calculations the **current datetime** instead of the **method parameter** was checked against these ranges to determine the current phase. 
This behavior is for almost all use-cases perfectly fine as the binding usually only shows the current sun phase. But when performing unit tests, this limits you to the current date and prevents you from checking historical phases etc. So, i changed it to the method parameter.

Also added three tests related to the issue. This revealed that the Map holding the ranges for all subphases is not sorted. giving strange results when null values get involved.

So, two fixes:
1. Fixed sorting of sun phases /ranges
2. Using SunPhase Calendar parameter instead of `Calendar.Instance` (now)

Closes https://github.com/openhab/openhab-addons/issues/7642
Closes https://github.com/openhab/openhab-addons/issues/12746
Closes https://github.com/openhab/openhab-addons/issues/12767

### CODE REVIEW

**Review by clinique — APPROVED** [review]

File: bundles/org.openhab.binding.astro/src/main/java/org/openhab/binding/astro/internal/model/MoonPhase.java [hunk 1]
```diff
@@ -85,14 +85,14 @@ public void setThirdQuarter(Calendar thirdQuarter) {
...
-        return _new;
+        return newCalendar;
     }
 
     /**
      * Sets the date of the new moon.
      */
     public void setNew(Calendar _new) {
-        this._new = _new;
+        this.newCalendar = _new;
```
Comment (lolodomo) [inline bundles/org.openhab.binding.astro/src/main/java/org/openhab/binding/astro/internal/model/MoonPhase.java:95]: You could also rename the parameter.

**Review by lolodomo — CHANGES_REQUESTED** [review]

**Review by lolodomo — APPROVED** [review]
LGTM, thank you

**Comment by lolodomo** [comment]
Is it something we should cherry-pick in 3.4.x ? 
@jlaur

**Comment by jlaur** [comment]
@lsiepel - good job hunting and nailing this long-time bug!

@lolodomo - I'm not sure. It's an equation of risk vs. impact. It seems this bug happens rarely (next time May?), but how sure are we that this fix will not cause any regressions? Perhaps we should leave it for 4.0 so we would have a full release cycle for detecting any regressions. I didn't look closely into the details, but please correct me if you believe the risk is negligible.

**Comment by lolodomo** [comment]
> It seems this bug happens rarely (next time May?), but how sure are we that this fix will not cause any regressions? Perhaps we should leave it for 4.0 so we would have a full release cycle for detecting any regressions.

Ok

**Comment by lsiepel** [comment]
> @lsiepel - good job hunting and nailing this long-time bug!
> 
> @lolodomo - I'm not sure. It's an equation of risk vs. impact. It seems this bug happens rarely (next time May?), but how sure are we that this fix will not cause any regressions? Perhaps we should leave it for 4.0 so we would have a full release cycle for detecting any regressions. I didn't look closely into the details, but please correct me if you believe the risk is negligible.

Agree, hope the milestones will get some feedback.

**Pull request merged**